lib,zebra,bgpd,ospfd,ospf6d: Yang and NB definitions for route-maps

NaveenThanikachalam opened the pull request:
This PR introduces the following changes:

Modifications to the existing library route-map yang definitions.
Introduction of yang definitions for bgpd, ospfd, ospf6d, and zebra route-map match and set clauses.
North-bound callbacks implementation for all the route-map yang elements.

Signed-off-by: NaveenThanikachalam <nthanikachal@vmware.com>

NaveenThanikachalam commented on bgpd/bgp_routemap.c:
```diff
@@ -3966,827 +3979,1140 @@ static void bgp_route_map_event(const char *rmap_name)
-       "Preference value (0-4294967295)\n")
+DEFUN(set_local_pref, set_local_pref_cmd, "set local-preference WORD",
+      SET_STR
+      "BGP local preference path attribute\n"
```
Thanks, Russ.  
I guess local-pref is still a path attribute for IBGP since it is transmitted to the internal BGP peers?

NaveenThanikachalam commented on bgpd/bgp_routemap.c:
```diff
@@ -3966,827 +3979,1140 @@ static void bgp_route_map_event(const char *rmap_name)
+DEFUN(no_set_local_pref, no_set_local_pref_cmd,
+      "no set local-preference [WORD]",
+      NO_STR SET_STR
+      "BGP local preference path attribute\n"
```
Thanks, Russ.  
I guess local-pref is still a path attribute for IBGP since it is transmitted to the internal BGP peers?

qlyoung commented on lib/routemap.h:
```diff
@@ -241,6 +241,146 @@ struct route_map {
 
+/* Route-map match conditions */
+#define IS_MATCH_INTERFACE(C)                                                  \
+	(strncmp(C, "frr-route-map:interface", strlen(C)) == 0)
```
More succinct to just convert all of these to
```
strmatch(C, "xpath")
```
I don't see the need for the case comparison as all xpaths should be lower case IMO

qlyoung commented on yang/frr-bgp-filter.yang:
```diff
@@ -0,0 +1,341 @@
+
+  typedef community-string {
+    type string {
+      pattern "(6553[0-5]|655[0-2][0-9]|65[0-4][0-9]{2}|6[0-4][0-9]{3}|[1-5][0-9]{4}|[1-9][0-9]{0,3}|0)|((6553[0-5]|655[0-2][0-9]|65[0-4][0-9]{2}|6[0-4][0-9]{3}|[1-5][0-9]{4}|[1-9][0-9]{0,3}|0):(6553[0-5]|655[0-2][0-9]|65[0-4][0-9]{2}|6[0-4][0-9]{3}|[1-5][0-9]{4}|[1-9][0-9]{0,3}|0))|(local-AS)|(no-advertise)|(no-export)|(internet)";
```
Should `additive` be one of the options here?

Also, why is `AS` capitalized in `local-AS`?

qlyoung commented on yang/frr-bgp-filter.yang:
```diff
@@ -0,0 +1,341 @@
+            "Extcommunity-list instance name/identifier";
+        }
+
+        choice Extcommunity-string {
```
This really should not be capitalized.

NaveenThanikachalam commented:
Thanks, @riw777, @glyoung, and @rzalamena for reviewing this PR.
I'll work on the comments this week.

dslicenc commented on yang/frr-bgp-filter.yang:
```diff
@@ -0,0 +1,341 @@
+          description
+            "Community-list instance name/identifier";
+        }
+
```
Why are there four options enumerated in large-community-lists and extcommunity-lists (standard and expanded ID and standard and expanded named) but only the two here combining the name/identifier?

dslicenc commented on yang/frr-bgp-filter.yang:
```diff
@@ -0,0 +1,341 @@
+          "AS-path access-list entry";
+        leaf sequence {
+          type list-sequence;
+        }
```
There isn't a sequence number on the vtysh command. Is this a unique requirement of the northbound/yang methodology?

NaveenThanikachalam commented:
Thanks, everyone for having an elaborate review/discussion on this.
I'll move this PR back into the draft state until I've addressed all the concerns around the changes.

donaldsharp commented:
I see a non-trivial amount of whitespace only changes that are causing problems allowing us to move this patch set forward.  I cannot seem to emphasize this enough..  Do Not Make WhiteSpace only changes to the code you submit.  It causes you and everyone else nothing but grief

Spantik commented:
closing this PR as https://github.com/FRRouting/frr/pull/7419 replaces it.
